# Chapter 2: Variables, Expressions, and Statements

## Section: Operators and Operands

### 🔣 What are Operators?

**Operators** are **special symbols** that represent computations like addition and multiplication.

Think of operators as **action symbols** that tell Python what to do:
- `+` means "add these together"
- `-` means "subtract this from that"
- `*` means "multiply these"

### 📦 What are Operands?

The **values** the operator is applied to are called **operands**.

**Example:**
```python
5 + 3
```
- `5` is an operand
- `+` is an operator
- `3` is an operand

**Analogy:** Like a sandwich:
- Bread (operand) + Peanut Butter (operator) + Bread (operand) = Sandwich!

In [1]:
# Simple example
result = 5 + 3
print(result)
print("In '5 + 3': 5 and 3 are operands, + is the operator")

8
In '5 + 3': 5 and 3 are operands, + is the operator


---
### ➕ Python's Arithmetic Operators

Python has several arithmetic operators for mathematical computations:

| Operator | Name | Example | Result |
|----------|------|---------|--------|
| `+` | Addition | `20 + 32` | `52` |
| `-` | Subtraction | `hour - 1` | Depends on hour |
| `*` | Multiplication | `5 * 3` | `15` |
| `/` | Division | `minute / 60` | Float result |
| `**` | Exponentiation (Power) | `5 ** 2` | `25` |
| `//` | Floor Division | `minute // 60` | Integer result |

### 📖 Examples from the Book

Let's try all the examples from the book:

In [2]:
# Example 1: Addition
print("20 + 32 =", 20 + 32)

20 + 32 = 52


In [3]:
# Example 2: Subtraction
hour = 10
print("hour - 1 =", hour - 1)

hour - 1 = 9


In [4]:
# Example 3: Multiplication and Addition combined
hour = 2
minute = 30
print("hour * 60 + minute =", hour * 60 + minute)
print("This converts 2 hours 30 minutes to total minutes:", hour * 60 + minute)

hour * 60 + minute = 150
This converts 2 hours 30 minutes to total minutes: 150


In [5]:
# Example 4: Division
minute = 59
print("minute / 60 =", minute / 60)
print("This converts minutes to hours (as decimal)")

minute / 60 = 0.9833333333333333
This converts minutes to hours (as decimal)


In [6]:
# Example 5: Exponentiation (Power)
print("5 ** 2 =", 5 ** 2)
print("This means 5 to the power of 2, or 5 × 5")

5 ** 2 = 25
This means 5 to the power of 2, or 5 × 5


In [7]:
# Example 6: Complex expression with parentheses
print("(5 + 9) * (15 - 7) =", (5 + 9) * (15 - 7))
print("First: 5 + 9 = 14")
print("Then: 15 - 7 = 8")
print("Finally: 14 * 8 = 112")

(5 + 9) * (15 - 7) = 112
First: 5 + 9 = 14
Then: 15 - 7 = 8
Finally: 14 * 8 = 112


---
### ⚠️ Important: Python 2 vs Python 3 Division

**There has been a change in the division operator between Python 2 and Python 3.**

#### In Python 3 (what you're using):
The `/` division operator returns a **floating-point result** (decimal)

In [9]:
# Python 3 division - returns float
minute = 59
result = minute / 60
print("Python 3: minute / 60 =", result)
print("Type:", type(result))  # It's a float!

Python 3: minute / 60 = 0.9833333333333333
Type: <class 'float'>


#### In Python 2 (older version):
Division of two integers would **truncate** (cut off) the decimal part:
```python
# Python 2 behavior (DON'T run this, just for reference)
>>> minute = 59
>>> minute / 60
0  # Lost the decimal part!
```

**Good news:** You're using Python 3, which works like a calculator! 🎉

### 🔢 Floor Division (//)

To get the **same behavior as Python 2** (integer result), use **floor division** (`//`):

**Floor division** divides and **rounds down** to the nearest integer.

In [10]:
# Floor division in Python 3
minute = 59
result = minute // 60
print("Python 3: minute // 60 =", result)
print("Type:", type(result))  # It's an integer!

Python 3: minute // 60 = 0
Type: <class 'int'>


### 📊 Comparison: / vs //

Let's see the difference side by side:

In [11]:
minute = 59

print("Regular division (/):", minute / 60)
print("Type:", type(minute / 60))
print()
print("Floor division (//):", minute // 60)
print("Type:", type(minute // 60))

Regular division (/): 0.9833333333333333
Type: <class 'float'>

Floor division (//): 0
Type: <class 'int'>


In [12]:
# More examples to see the difference
print("=== Regular Division (/) ===")
print("10 / 3 =", 10 / 3)
print("15 / 2 =", 15 / 2)
print("7 / 2 =", 7 / 2)
print()
print("=== Floor Division (//) ===")
print("10 // 3 =", 10 // 3)
print("15 // 2 =", 15 // 2)
print("7 // 2 =", 7 // 2)

=== Regular Division (/) ===
10 / 3 = 3.3333333333333335
15 / 2 = 7.5
7 / 2 = 3.5

=== Floor Division (//) ===
10 // 3 = 3
15 // 2 = 7
7 // 2 = 3


### 🧮 Python 3 Works Like a Calculator!

**Key Point:** In Python 3, integer division functions much more as you would expect if you entered the expression on a calculator.

This is a **good thing!** It's more intuitive.

In [13]:
# Calculator-like behavior
print("On a calculator: 59 ÷ 60 =", 59 / 60)
print("Python 3 matches this! ✅")

On a calculator: 59 ÷ 60 = 0.9833333333333333
Python 3 matches this! ✅


---
### 🔍 Understanding Each Operator in Detail

#### 1. Addition (+)

In [14]:
# Addition examples
print("5 + 3 =", 5 + 3)
print("100 + 50 =", 100 + 50)
print("1.5 + 2.5 =", 1.5 + 2.5)

# With variables
a = 10
b = 20
print("a + b =", a + b)

5 + 3 = 8
100 + 50 = 150
1.5 + 2.5 = 4.0
a + b = 30


#### 2. Subtraction (-)

In [15]:
# Subtraction examples
print("10 - 3 =", 10 - 3)
print("100 - 25 =", 100 - 25)
print("5.5 - 2.5 =", 5.5 - 2.5)

# Can result in negative numbers
print("3 - 10 =", 3 - 10)

10 - 3 = 7
100 - 25 = 75
5.5 - 2.5 = 3.0
3 - 10 = -7


#### 3. Multiplication (*)

In [16]:
# Multiplication examples
print("5 * 3 =", 5 * 3)
print("12 * 12 =", 12 * 12)
print("2.5 * 4 =", 2.5 * 4)

# Useful for conversions
hours = 3
minutes_per_hour = 60
total_minutes = hours * minutes_per_hour
print(f"{hours} hours = {total_minutes} minutes")

5 * 3 = 15
12 * 12 = 144
2.5 * 4 = 10.0
3 hours = 180 minutes


#### 4. Division (/)

**Always returns a float in Python 3!**

In [17]:
# Division examples
print("10 / 2 =", 10 / 2)      # Even division still gives float!
print("15 / 4 =", 15 / 4)
print("100 / 3 =", 100 / 3)

# Always float, even with whole number result
print("Type of 10 / 2:", type(10 / 2))

10 / 2 = 5.0
15 / 4 = 3.75
100 / 3 = 33.333333333333336
Type of 10 / 2: <class 'float'>


#### 5. Floor Division (//)

**Divides and rounds DOWN to nearest integer**

In [18]:
# Floor division examples
print("10 // 3 =", 10 // 3)    # 3.33... rounds down to 3
print("15 // 4 =", 15 // 4)    # 3.75 rounds down to 3
print("20 // 6 =", 20 // 6)    # 3.33... rounds down to 3
print("7 // 2 =", 7 // 2)      # 3.5 rounds down to 3

# Useful for: "How many complete groups?"
students = 25
students_per_group = 4
complete_groups = students // students_per_group
print(f"With {students} students, we can make {complete_groups} complete groups of {students_per_group}")

10 // 3 = 3
15 // 4 = 3
20 // 6 = 3
7 // 2 = 3
With 25 students, we can make 6 complete groups of 4


#### 6. Exponentiation (**)

**Raises a number to a power**

In [19]:
# Exponentiation examples
print("2 ** 3 =", 2 ** 3)      # 2 × 2 × 2 = 8
print("5 ** 2 =", 5 ** 2)      # 5 × 5 = 25
print("10 ** 3 =", 10 ** 3)    # 10 × 10 × 10 = 1000
print("2 ** 8 =", 2 ** 8)      # 2 × 2 × 2 × 2 × 2 × 2 × 2 × 2 = 256

# Works with floats too
print("2 ** 0.5 =", 2 ** 0.5)  # Square root of 2!
print("9 ** 0.5 =", 9 ** 0.5)  # Square root of 9 = 3

2 ** 3 = 8
5 ** 2 = 25
10 ** 3 = 1000
2 ** 8 = 256
2 ** 0.5 = 1.4142135623730951
9 ** 0.5 = 3.0


### 🎯 Bonus: Modulo Operator (%)

While not in the book section, this is another useful operator:

**Modulo (%)** returns the **remainder** after division

In [20]:
# Modulo examples
print("10 % 3 =", 10 % 3)      # 10 ÷ 3 = 3 remainder 1
print("15 % 4 =", 15 % 4)      # 15 ÷ 4 = 3 remainder 3
print("20 % 5 =", 20 % 5)      # 20 ÷ 5 = 4 remainder 0 (divides evenly)

# Useful for: "How many are left over?"
students = 25
students_per_group = 4
leftover_students = students % students_per_group
print(f"With {students} students in groups of {students_per_group}, {leftover_students} students are left over")

10 % 3 = 1
15 % 4 = 3
20 % 5 = 0
With 25 students in groups of 4, 1 students are left over


---
### 📝 Practice Exercises

In [ ]:
# Exercise 1: Basic operations
# Calculate and print:
# a) 15 + 27
# b) 100 - 37
# c) 8 * 9
# d) 50 / 4
# e) 3 ** 4

# Your code here:


In [ ]:
# Exercise 2: Division comparison
# For the number 47:
# Calculate 47 / 5 (regular division)
# Calculate 47 // 5 (floor division)
# Calculate 47 % 5 (modulo - remainder)
# Print all three results with labels

# Your code here:


In [ ]:
# Exercise 3: Time conversion
# You have 150 minutes
# Convert to hours and remaining minutes
# Hint: Use // for hours, % for remaining minutes

total_minutes = 150
# Your code here:


In [ ]:
# Exercise 4: Calculate area and perimeter
# Rectangle with length = 12, width = 5
# Area = length * width
# Perimeter = 2 * (length + width)

length = 12
width = 5
# Your code here:


In [ ]:
# Exercise 5: Exponentiation practice
# Calculate:
# a) 2 to the power of 10
# b) 3 squared (3 to the power of 2)
# c) 5 cubed (5 to the power of 3)

# Your code here:


### 🎯 Challenge Exercises

In [ ]:
# Challenge 1: Temperature conversion
# Convert Celsius to Fahrenheit
# Formula: F = (C * 9/5) + 32
# Convert 25°C to Fahrenheit

celsius = 25
# Your code here:


In [ ]:
# Challenge 2: Circle calculations
# Given radius = 7
# Calculate area = π * r²
# Calculate circumference = 2 * π * r
# Use 3.14159 for π

radius = 7
pi = 3.14159
# Your code here:


In [ ]:
# Challenge 3: Splitting money
# You have $100 to split among 7 people
# How much does each person get? (use /)
# How many complete dollars? (use //)
# How many cents left over? (use %)

total_money = 100
people = 7
# Your code here:


---
### 📊 Quick Reference Table

| Operator | Name | Example | Result | Type |
|----------|------|---------|--------|------|
| `+` | Addition | `5 + 3` | `8` | Same as operands |
| `-` | Subtraction | `10 - 3` | `7` | Same as operands |
| `*` | Multiplication | `4 * 5` | `20` | Same as operands |
| `/` | Division | `10 / 3` | `3.333...` | **Always float** |
| `//` | Floor Division | `10 // 3` | `3` | Integer |
| `%` | Modulo | `10 % 3` | `1` | Integer |
| `**` | Exponentiation | `2 ** 3` | `8` | Depends on operands |

### 📌 Key Takeaways

- ✅ **Operators** are special symbols for computations (+, -, *, /, **, //)
- ✅ **Operands** are the values operators work on
- ✅ Six main arithmetic operators: `+`, `-`, `*`, `/`, `**`, `//`
- ✅ In **Python 3**, `/` always returns a **float** (decimal)
- ✅ Use `//` for **floor division** (rounds down to integer)
- ✅ Python 3 division works **like a calculator** (intuitive!)
- ✅ `**` is for **exponentiation** (powers)
- ✅ Python 3 is **different from Python 2** for division
- ✅ `%` (modulo) gives the **remainder** (bonus operator!)

**Remember:** `/` gives float, `//` gives integer (rounded down)

### 🤔 Quick Quiz

1. What's the difference between `10 / 3` and `10 // 3`?
2. What does `5 ** 2` calculate?
3. In `a + b`, what are `a` and `b` called?
4. Does Python 3 division work like a calculator? Why is this good?
5. What operator would you use to find "how many complete groups"?

**Your answers:**

1. 
2. 
3. 
4. 
5. 

### 💭 My Notes

*Important points or questions:*

- 
- 
- 

---
### 🎯 Next Steps

Now that you understand operators and operands, you're ready to learn:
- **Expressions** - Combining operators and operands
- **Order of operations** - Which operations happen first (PEMDAS)
- **More complex calculations**

**You're making excellent progress!** 🚀